jaxlib/rocblas.cc

```diff
@@ -0,0 +1,901 @@
+/* Copyright 2019 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <algorithm>
+#include <stdexcept>
+#include <utility>
+#include <vector>
+
+
+#include "absl/base/casts.h"
+#include "absl/base/thread_annotations.h"
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/str_format.h"
+#include "absl/synchronization/mutex.h"
+#include "include/pybind11/numpy.h"
+#include "include/pybind11/pybind11.h"
+#include "include/pybind11/stl.h"
+#include "jaxlib/handle_pool.h"
+#include "jaxlib/kernel_pybind11_helpers.h"
+#include "jaxlib/rocm_gpu_kernel_helpers.h"
+#include "rocm/include/hip/hip_runtime.h"
+#include "rocm/include/hip/hip_runtime_api.h"
+#include "rocm/include/rocblas.h"
+#include "rocm/include/rocsolver.h"
+
+
+namespace jax {
+namespace {
+
+namespace py = pybind11;
+
+void ThrowIfErrorStatus(rocblas_status status) {
+  switch (status) {
+    case rocblas_status_success:
+      return;
+    default:
+      throw std::runtime_error(rocblas_status_to_string(status));
+  }
+}
+
+using rocBlasHandlePool = HandlePool<rocblas_handle, hipStream_t>;
+
+template <>
+/*static*/ rocBlasHandlePool::Handle rocBlasHandlePool::Borrow(
+    hipStream_t stream) {
+  rocBlasHandlePool* pool = Instance();
+  absl::MutexLock lock(&pool->mu_);
+  rocblas_handle handle;
+  if (pool->handles_.empty()) {
+    ThrowIfErrorStatus(rocblas_create_handle(&handle));
```

Comment:
@hawkinsp Here a handle is created, but we never destroy it (same for cuda).

Reply:
I think this code is correct: we pass ownership of the handle to the `::Handle` class. When the `Handle` is released, ownership is returned to the pool.

We never shrink the pool, so there's no need to release the underlying handles.

We could, I guess, delete the handles in the pool destructor, but note the pool is a global singleton that is itself never deallocated, so it wouldn't change anything.